src/main/java/swinglib/ErrorMessageBuilder.java

```diff
@@ -0,0 +1,123 @@
+
+
+/**
+ * Swing builder for an error message pop-up (Uses JOptionPane). Example usages:
+ * <code><pre>
+ *   JOptionPaneBuilder.builder()
+ *     .title("Something Crashed")
+ *     .message("Error Details")
+ *     .option("ok", () -> {})
+ *     .option("quit", () -> System.exit(0))
+ *     .buildAndShow();
+ * </pre></code>
+ * Default error dialog with an 'ok' button:
+ * <code><pre>
+ *   JOptionPaneBuilder.builder()
+ *     .message("error details")
+ *     .buildAndShow();
+ * </pre></code>
+ */
+public class ErrorMessageBuilder {
+
+  private String title = "Error";
+  private String message;
+  private List<CloseButton> actions = new ArrayList<>();
+
+  private static class CloseButton {
+    private final String buttonText;
+    private final Runnable closeAction;
+
+    private CloseButton(String buttonText, Runnable closeAction) {
+      this.buttonText = buttonText;
+      this.closeAction = closeAction;
+    }
+  }
+
+  private ErrorMessageBuilder() {}
+
+  public static ErrorMessageBuilder builder() {
+    return new ErrorMessageBuilder();
+  }
+
+  public ErrorMessageBuilder message(String message) {
+    this.message = message;
+    return this;
+  }
+
+  /**
+   * Constructs a Swing JDialog using current builder values.
+   * Values that must be set: title, contents
+   */
+  public void buildAndShow() {
+    checkState(!GraphicsEnvironment.isHeadless(), "JOptionPane throws an exception in headless environments");
+    checkNotNull(message);
+
+    if (actions.isEmpty()) {
+      actions.add(new CloseButton("OK", () -> {
+      }));
+    }
+
+    // First option element to JOptionPane is the right most button, right to left ordering.
+    // It is more natural for left-to-right ordering of options to be specified, so we
+    // will reverse the option ordering to allow clients to specify left to right ordering.
+
+    List<CloseButton> swingOptionsRightToLeft = reverse(actions);
+
+    String[] buttonOptions = swingOptionsRightToLeft.stream()
+        .map(button -> button.buttonText)
+        .toArray(String[]::new);
+
+    int value = JOptionPane.showOptionDialog(
+        null,
+        message,
+        title,
+        JOptionPane.DEFAULT_OPTION,
+        JOptionPane.ERROR_MESSAGE,
+        null,
+        buttonOptions,
+        0);
+
+    swingOptionsRightToLeft.get(value).closeAction.run();
+
+  }
+
+
+  private static List<CloseButton> reverse(List<CloseButton> input) {
+    List<CloseButton> output = new ArrayList<>();
+    input.forEach(element -> output.add(0, element));
+    return output;
+  }
+
+  /**
+   * Adds another 'option' button to error message dialog window. Buttons are added in left to right order.
+   * 
+   * @param buttonText Text of the option button.
+   * @param buttonAction Action to be performed when the button is clicked.
+   */
+  public ErrorMessageBuilder option(final String buttonText, Runnable buttonAction) {
+    actions.add(new CloseButton(buttonText, buttonAction));
```

Comment:
If I didn't explain it well earlier, if this line would be replaced with `actions.add(0,...)` There wouldn't be a need to resort the list when building.
Not sure how well the performance is for this though.

Reply:
Perf is not an issue, we expect 2 or 3 elements, more than 3 buttons on an error dialog seems like too many options.

I now understand the suggestion, thanks! 602c524e3 incorporates it, very early on in the lift of this branch I gave something like that a try, for some I did not get it to work and  hacked out a 'reverse' method instead. Removing the 'reverse' method is for sure a win.